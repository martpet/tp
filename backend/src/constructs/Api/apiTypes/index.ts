export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
