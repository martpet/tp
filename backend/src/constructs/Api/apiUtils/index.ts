export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
