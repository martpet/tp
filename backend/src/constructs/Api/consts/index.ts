export * from './authorizationHeader';
export * from './defaultCachePolicyProps';
export * from './defaultSecurityHeadersBehavior';
export * from './lambdaEdgeViewerEvent';
