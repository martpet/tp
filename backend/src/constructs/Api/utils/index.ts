export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
