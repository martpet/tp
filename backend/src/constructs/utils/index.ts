export * from './createEdgeFunction';
export * from './createNodejsFunction';
