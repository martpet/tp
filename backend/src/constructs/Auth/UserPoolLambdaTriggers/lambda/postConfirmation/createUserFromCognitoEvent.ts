import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { PostConfirmationTriggerEvent } from 'aws-lambda';

import { usersTableOptions } from '~/consts';
import { UsersTableItem } from '~/types';

import { getUserPropsFromCognitoEvent } from '../getUserPropsFromCognitoEvent';

const marshallOptions = { removeUndefinedValues: true };
const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient, { marshallOptions });

export const createUserFromCognitoEvent = (event: PostConfirmationTriggerEvent) => {
  const item: UsersTableItem = {
    ...getUserPropsFromCognitoEvent(event),
    settings: {},
  };

  const putCommand = new PutCommand({
    TableName: usersTableOptions.tableName,
    Item: item,
  });

  return ddbDocClient.send(putCommand);
};
