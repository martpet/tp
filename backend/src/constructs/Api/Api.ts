import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/authEdgeFunction/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'HttpApi');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'Alias', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
