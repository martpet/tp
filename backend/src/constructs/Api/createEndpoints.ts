import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import {
  CachePolicy,
  Distribution,
  OriginRequestCookieBehavior,
  OriginRequestHeaderBehavior,
  OriginRequestPolicy,
  OriginRequestPolicyProps,
  OriginRequestQueryStringBehavior,
} from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { Writable } from 'type-fest';

import { capitalize, createNodejsFunction } from '~/utils';

import { EndpointsOptions, MethodOptions } from './apiTypes';

type CreateEndpointProps = {
  scope: Construct;
  distribution: Distribution;
  api: HttpApi;
  origin: HttpOrigin;
  options: EndpointsOptions;
};

export const createEndpoints = ({
  scope,
  distribution,
  api,
  origin,
  options,
}: CreateEndpointProps) => {
  Object.entries(options).forEach(
    ([path, { methods, cookies, headers, queryStrings }]) => {
      addDistroBehavior({
        scope,
        distribution,
        origin,
        path,
        headers,
        cookies,
        queryStrings,
      });

      Object.entries(methods).forEach(([method, methodOptions]) => {
        addApiRoute({ scope, api, path, method, methodOptions });
      });
    }
  );
};

type AddDistroBehaviorProps = {
  scope: Construct;
  distribution: Distribution;
  origin: HttpOrigin;
  path: string;
  headers?: string[];
  cookies?: string[];
  queryStrings?: Readonly<string[]>;
};

function addDistroBehavior({
  scope,
  distribution,
  origin,
  path,
  cookies,
  headers,
  queryStrings,
}: AddDistroBehaviorProps) {
  const pathName = path.replace('/', '');
  const policyProps: Writable<OriginRequestPolicyProps> = {};

  if (headers) {
    policyProps.headerBehavior = OriginRequestHeaderBehavior.allowList(...headers);
  }

  if (cookies) {
    policyProps.cookieBehavior = OriginRequestCookieBehavior.allowList(...cookies);
  }

  if (queryStrings) {
    policyProps.queryStringBehavior = OriginRequestQueryStringBehavior.allowList(
      ...queryStrings
    );
  }

  distribution.addBehavior(path, origin, {
    cachePolicy: CachePolicy.CACHING_DISABLED,
    originRequestPolicy: new OriginRequestPolicy(scope, pathName, policyProps),
  });
}

type AddApiRouteProps = {
  scope: Construct;
  api: HttpApi;
  path: string;
  method: string;
  methodOptions: MethodOptions<unknown>;
};

function addApiRoute({ scope, api, path, method, methodOptions }: AddApiRouteProps) {
  const pathName = path.replace('/', '');
  const methodName = method.toLowerCase();
  const fileName = `${methodName}-${pathName}`;
  const id = `${capitalize(methodName)}${capitalize(fileName)}`;
  const { environment, cb } = methodOptions;
  const handlerProps: Writable<NodejsFunctionProps> = {
    entry: `${__dirname}/handlers/${pathName}/${fileName}/${fileName}.ts`,
    environment,
  };

  const handler = createNodejsFunction(scope, `${id}Handler`, handlerProps);

  if (cb) {
    cb(handler);
  }

  api.addRoutes({
    path,
    methods: [method as HttpMethod],
    integration: new HttpLambdaIntegration(`${id}Integration`, handler),
  });
}
