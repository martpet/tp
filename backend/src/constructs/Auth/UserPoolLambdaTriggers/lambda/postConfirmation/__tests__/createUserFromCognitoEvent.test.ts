import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { itResolves, itSendsDdbCommand } from '~/constructs/Api/utils';

import { createUserFromCognitoEvent } from '../createUserFromCognitoEvent';
import event from './__fixtures__/postConfirmationEvent';

vi.mock('../../getUserPropsFromCognitoEvent');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [event] as Parameters<typeof createUserFromCognitoEvent>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock
    .on(PutCommand)
    .resolves({ Attributes: { dummyPutCommandAttrKey: 'dummyPutCommandAttrValue' } });
});

describe('createUserFromCognitoEvent', () => {
  itSendsDdbCommand(PutCommand, ddbMock, createUserFromCognitoEvent, args);

  itResolves(createUserFromCognitoEvent, args);
});
