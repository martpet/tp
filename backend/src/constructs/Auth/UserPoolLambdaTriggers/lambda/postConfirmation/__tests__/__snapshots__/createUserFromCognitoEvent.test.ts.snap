// Vitest Snapshot v1

exports[`createUserFromCognitoEvent > resolves with a correct value 1`] = `
{
  "Attributes": {
    "dummyPutCommandAttrKey": "dummyPutCommandAttrValue",
  },
}
`;

exports[`createUserFromCognitoEvent > sends "PutCommand" to DynamoDB with correct args 1`] = `
{
  "Item": {
    "dummyUserPropFromCognitoEventKey": "dummyUserPropFromCognitoEventValue",
    "id": "dummyId",
    "settings": {},
  },
  "TableName": "TripPics-Users",
}
`;
